### ISSUE TITLE
angular 2 ng2-smart table: Cannot read property 'subscribe' of undefined

### ISSUE THREAD

**Issue by Sachin-Suresh** [akveo/ng2-smart-table#470]
Im trying to display button in ng2-smart table, both on click of 'edit' link and on page load for which my column settings is as:

**html**

    <ng2-smart-table [settings]="view_update_items_settings" [source]="source" (createConfirm)="onCreateConfirm($event)" (editConfirm)="onSaveConfirm($event)"  (deleteConfirm)="onDeleteConfirm($event)" class="table table-hover table-striped"></ng2-smart-table>

**component**

     //Confirm Create before save to table function
      onCreateConfirm(event) {
    if (window.confirm('Are you sure you want to create?')) {
      
      event.newData.product_image =  localStorage.getItem('image_name');
     
      this.productService.productSave(event.newData, this.selectedObject).then(data => {
      
        this.flashMessage.show('Product successfully added', { cssClass: 'alert-success', timeout: 3000 });
        this.router.navigate(['/category-items']);
      }, (err) => {
      
        this.flashMessage.show('Something went wrong', { cssClass: 'alert-danger', timeout: 3000 });
        this.router.navigate(['/category-items']);

      });
      event.confirm.resolve(event.newData);
    } else {
      
      event.confirm.reject();
         }
      }
    
    view_update_items_settings = {
    add: {
      confirmCreate: true,
    },
    edit: {
      confirmSave: true,
    },
    delete: {
      confirmDelete: true,
    },
    columns: {
      product_name: {
        title: 'Product Name'
      },
      product_price: {
        title: 'Product Price'
      },
      product_image: {
        title: 'Product Image',
        type: 'custom',
        renderComponent: ButtonViewComponent,
        editor: {
         type: 'custom',
          component: ButtonViewComponent,  //To get Image upload button on create
        },
      },
      product_desc: {
        title: 'Product Desc'
      },
      product_status: {
        title: 'Status',
        filter: {
          type: 'list',
          config: {
            selectText: 'Select',
            list: [
              { value: 'Yes', title: 'Yes' },
              { value: 'No', title: 'No' }
            ],
          },
        }, //filter close
      },
    },
      };

**productservice**

     //Save new Product
     productSave(products, categoryId) {

    let prodDet = {
      catId: categoryId,
      productDetail: products
    }

    let headers = new Headers();
    headers.append('Content-Type', 'application/json');
    return new Promise((resolve, reject) => {
      this.http.post(this.domain + 'products/add', { prodDet }, { headers: headers })
        .map(res => res.json())
        .subscribe(res => {
          resolve(res);
        }, (err) => {
          reject(err);
        });
    });

      }

But when i click on 'Add new' or 'edit'  link from the ng2-smart-table, i get 

> Cannot read property 'subscribe' of undefined

**Error stack trace:**

    ERROR TypeError: Cannot read property 'subscribe' of undefined
    at CustomEditComponent.webpackJsonp.../../../../ng2-smart-table/components/cell/cell-edit-mode/custom-edit.component.js.CustomEditComponent.ngOnChanges

**Comment by digiphaze** [comment]
Map isn't a default of @angular2,  have you made sure to import it from rxjs?
`import 'rxjs/add/operator/map';`

**Comment by Sachin-Suresh** [comment]
Yes I have imported rxjs map operator...But still I get the message in the console, even though functionality works... 😢

**Comment by dvirHam** [comment]
I have a similar situation. It seems there is couple of EventEmitter-s which on ngOnChange it subscribes from them, which it can't be done since they are'nt observables. Look like it's a bug.

**Comment by Requium** [comment]
Im not an expert on the subject, but i follow the example https://akveo.github.io/ng2-smart-table/#/examples/custom-editors-viewers (Custom editor/render Column Type), and i noticed this code:
`export class CustomEditorComponent **extends DefaultEditor** implements AfterViewInit {
` 
and this:
` constructor() {
    **super();**
  }`
Adding that to the component you are calling from the editor will allow you to avoid that error.

by the way you have to import Default editor from ng2-smart table on the component you are calling:
`import { Cell, DefaultEditor, Editor} from 'ng2-smart-table';`

Like i said im not expert on angular or on the subject, i cant tell you why you have to put that, i just copied from the example, maybe someone can clarify that process to us hehe.

**Comment by vdanniel** [comment]
that works. use of @viewchild and extends DefaultEditor implementing AfterViewInit

**Comment by hope4555** [comment]
happens if you use a renderer instead of an editor
